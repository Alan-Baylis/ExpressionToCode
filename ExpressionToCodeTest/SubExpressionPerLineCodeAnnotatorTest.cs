using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ExpressionToCodeLib;
using Xunit;

namespace ExpressionToCodeTest
{
    public class SubExpressionPerLineCodeAnnotatorTest
    {
        static readonly ExpressionToCodeConfiguration config = ExpressionToCodeConfiguration.DefaultAssertionConfiguration.WithAnnotator(CodeAnnotators.SubExpressionPerLineCodeAnnotator);
        static readonly IAnnotatedToCode annotator = config.GetAnnotatedToCode();

        [Fact]
        public void AplusBapproved()
        {
            var a = 2;
            var b = 5;
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a + b));
        }

        [Fact]
        public void NestedArrayAccess()
        {
            var a = 2;
            var b = 5;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => nums[a + b] < 7));
        }

        [Fact]
        public void NestedArrayAccessWithOuterAnd()
        {
            var a = 2;
            var b = 5;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a < b && nums[a + b] < 7 && b < 10));
        }

        [Fact]
        public void Binary_expressions_with_nesting()
        {
            var a = 2;
            var b = 5;
            var c = 3.45;
            var nums = Enumerable.Range(10, 10).ToArray();
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => a < b && (c > -a || c > b) && b < 10));
        }

        [Fact]
        public void MethodCallsAndArrayLiterals()
        {
            var a = 2;
            var b = 5;
            ApprovalTest.Verify(annotator.AnnotatedToCode(() => Math.Max(a, b) > new[] { 3, 8, 13, 4 }.Average()));
        }

        [Fact]
        public void MessyStructureElidesNeatly() {
            var hmm = "1234567890";
            ApprovalTest.Verify(annotator.AnnotatedToCode(
                () => hmm[1] == hmm[2] || hmm[4] == hmm[int.Parse(hmm[8].ToString())] || false
                ));
        }
    }
}
