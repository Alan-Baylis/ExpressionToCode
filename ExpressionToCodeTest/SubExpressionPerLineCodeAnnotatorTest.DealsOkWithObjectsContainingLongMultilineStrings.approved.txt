assertion failed

new { A_wall_of_text = wallOfText, A_short_string = "short", A_long_enumerable = Enumerable.Range(0, 1000) }
   →   @"line 0:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      1 ...
          line 1:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 2:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 3:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 4:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 5:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 6:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 7:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 8:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 9:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 10:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 11:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 12:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 13:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 14:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 15:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 16:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 17:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 18:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 19:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 20:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 21:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 22:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 23:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 24:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 25:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 26:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 27:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 28:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 29:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          ... (caused assertion failure)

wallOfText
     →   @"line 0:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      1 ...
          line 1:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 2:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 3:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 4:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 5:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 6:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 7:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 8:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 9:          20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 10:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 11:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 12:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 13:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 14:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 15:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 16:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 17:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 18:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 19:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 20:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 21:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 22:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 23:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 24:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 25:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 26:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 27:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 28:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          line 29:         20;       30;       40;       50;       60;       70;       80;       90;      100;      110;      120;      130;      140;      150 ...
          ...

Enumerable.Range(0, 1000)
     →   { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, ... }

